# Collecting the Data

In [1]:
import pandas as pd

### 1 Phishing URLs

In [2]:
!wget http://data.phishtank.com/data/online-valid.csv

'wget' is not recognized as an internal or external command,
operable program or batch file.


In [3]:
phishing_data = pd.read_csv("online-valid.csv")
phishing_data

phish_id                                                url  \
0       8134350          http://walleterrorcorrectionweb.pages.dev   
1       8134348  http://pagebussinessreview.strangled.net/?pgvi...   
2       8134347          http://pagebussinessreview.strangled.net/   
3       8134345                      http://collabauth.netlify.app   
4       8134340  https://staff.fatturazione-elettronica.net/Admin/   
...         ...                                                ...   
61463   1865971  http://www.formbuddy.com/cgi-bin/formdisp.pl?u...   
61464   1850379      https://sites.google.com/site/libretyreserve/   
61465   1460953       http://www.habbocreditosparati.blogspot.com/   
61466   1123978  http://creditiperhabbogratissicuro100.blogspot...   
61467    549159  http://aijcs.blogspot.com/2005/03/colourful-li...   

                                        phish_detail_url  \
0      http://www.phishtank.com/phish_detail.php?phis...   
1      http://www.phishtank.com/phish_detail.php?phis...   
2      http://www.phishtank.com/phish_detail.php?phis...   
3      http://www.phishtank.com/phish_detail.php?phis...   
4      http://www.phishtank.com/phish_detail.php?phis...   
...                                                  ...   
61463  http://www.phishtank.com/phish_detail.php?phis...   
61464  http://www.phishtank.com/phish_detail.php?phis...   
61465  http://www.phishtank.com/phish_detail.php?phis...   
61466  http://www.phishtank.com/phish_detail.php?phis...   
61467  http://www.phishtank.com/phish_detail.php?phis...   

                 submission_time verified          verification_time online  \
0      2023-04-30T07:51:46+00:00      yes  2023-04-30T07:52:54+00:00    yes   
1      2023-04-30T07:44:42+00:00      yes  2023-04-30T07:52:54+00:00    yes   
2      2023-04-30T07:44:03+00:00      yes  2023-04-30T07:52:54+00:00    yes   
3      2023-04-30T07:44:00+00:00      yes  2023-04-30T07:52:54+00:00    yes   
4      2023-04-30T07:40:02+00:00      yes  2023-04-30T07:43:28+00:00    yes   
...                          ...      ...                        ...    ...   
61463  2013-05-28T15:59:31+00:00      yes  2013-05-28T22:13:12+00:00    yes   
61464  2013-05-18T23:52:56+00:00      yes  2013-05-19T13:45:13+00:00    yes   
61465  2012-06-14T15:19:26+00:00      yes  2012-06-14T15:29:11+00:00    yes   
61466  2011-02-18T10:50:32+00:00      yes  2011-02-19T00:05:43+00:00    yes   
61467  2008-11-07T15:04:00+00:00      yes  2011-09-03T19:15:33+00:00    yes   

                   target  
0                   Other  
1                Facebook  
2                Facebook  
3                   Other  
4                   Other  
...                   ...  
61463               Other  
61464               Other  
61465  Sulake Corporation  
61466  Sulake Corporation  
61467               Other  

[61468 rows x 8 columns]

In [4]:
phishing_data.shape

(61468, 8)

Here we have 61468 phishing URLs. We will pick 1400 samples randomly for our dataset

In [45]:
phishingURL = phishing_data.sample(n = 1400, random_state = 12).copy()
phishingURL = phishingURL.reset_index(drop = True)
phishingURL

phish_id                                                url  \
0      8012441  http://www-nova-agora-show.com/589458/002366r/...   
1      8133956                      http://mszhrgkpuy.duckdns.org   
2      7850567  https://infocantik.my.id/mastah-kampret/0lmLz9...   
3      8064680     https://www.wnsqyr.com/sanjin/client/index.php   
4      8051492               https://prestazioni-id.info/account/   
...        ...                                                ...   
1395   8068635  https://www.jqincould.com/sanjin/client/index.php   
1396   8027060         http://help-center262.crabdance.com/verify   
1397   8116423               https://qrs.ly/qdeqh7k?mekicinasdfsd   
1398   8012650  https://novaconsultamagalu.com/consulte-sua-fa...   
1399   8080299  https://steampay.cc/steam/store.steampowered.com/   

                                       phish_detail_url  \
0     http://www.phishtank.com/phish_detail.php?phis...   
1     http://www.phishtank.com/phish_detail.php?phis...   
2     http://www.phishtank.com/phish_detail.php?phis...   
3     http://www.phishtank.com/phish_detail.php?phis...   
4     http://www.phishtank.com/phish_detail.php?phis...   
...                                                 ...   
1395  http://www.phishtank.com/phish_detail.php?phis...   
1396  http://www.phishtank.com/phish_detail.php?phis...   
1397  http://www.phishtank.com/phish_detail.php?phis...   
1398  http://www.phishtank.com/phish_detail.php?phis...   
1399  http://www.phishtank.com/phish_detail.php?phis...   

                submission_time verified          verification_time online  \
0     2023-01-21T01:32:08+00:00      yes  2023-01-21T01:42:05+00:00    yes   
1     2023-04-30T00:17:58+00:00      yes  2023-04-30T00:25:43+00:00    yes   
2     2022-10-17T06:52:26+00:00      yes  2022-10-17T10:53:13+00:00    yes   
3     2023-03-07T16:04:55+00:00      yes  2023-03-07T16:35:00+00:00    yes   
4     2023-02-27T12:08:20+00:00      yes  2023-02-27T12:12:00+00:00    yes   
...                         ...      ...                        ...    ...   
1395  2023-03-10T15:20:01+00:00      yes  2023-03-10T15:35:18+00:00    yes   
1396  2023-02-06T01:20:35+00:00      yes  2023-02-06T01:32:40+00:00    yes   
1397  2023-04-14T00:38:34+00:00      yes  2023-04-14T00:55:13+00:00    yes   
1398  2023-01-21T15:12:09+00:00      yes  2023-01-21T15:22:07+00:00    yes   
1399  2023-03-18T00:33:21+00:00      yes  2023-03-18T01:35:15+00:00    yes   

                                   target  
0                                   Other  
1                                   Apple  
2                                   Other  
3     Sumitomo Mitsui Banking Corporation  
4                                   Other  
...                                   ...  
1395  Sumitomo Mitsui Banking Corporation  
1396                                Other  
1397                                Other  
1398                                Other  
1399                                Other  

[1400 rows x 8 columns]

In [46]:
phishingURL.shape

(1400, 8)

Now, we have extracted the 1400 phishing URLs which would be used for training the model.

### 2 Legitimate URLs

In [47]:
legitimate_data = pd.read_csv("legitimate-urls.csv")
legitimate_data.columns = ['URLs']
legitimate_data

URLs
0     https://www.camosy.com/themes/juicy/images/doo...
1                          https://www.liveapps.com.au/
2                            https://www.oec.co.jp/csr/
3                       https://www.stable.cz/napoveda/
4           https://www.tektonic.net/privacypolicy.html
...                                                 ...
2948  https://www.redytel.info/wp-content/uploads/fb...
2949  https://www.ehesp.fr/2021/03/26/9eme-conferenc...
2950        https://www.qsat.net/managed-it-networking/
2951           https://www.edpa.com/trade-show-calendar
2952                     https://www.heritage-house.eu/

[2953 rows x 1 columns]

In [48]:
legitimate_data.shape

(2953, 1)

Here we have 2953 Legitimate URLs. Out of these 2953 URLs, we would pick 1100 random URLs for our training purpose.

In [49]:
legitimateURL = legitimate_data.sample(n = 1100, random_state = 12).copy()
legitimateURL = legitimateURL.reset_index(drop = True)
legitimateURL

URLs
0               https://www.bilkom.com.tr/apple-ailesi/
1           https://www.cognatebioservices.com/about-us
2     https://www.xelleration.com/resources/case-stu...
3     https://www.mawa-solutions.com/produkte/mawadc...
4                    https://www.primuselectronics.com/
...                                                 ...
1095               https://www.furlongsteel.ca/wp-json/
1096      https://www.innandsuitesatriverwalk.com/feed/
1097                        https://www.toc.lt/mokymai/
1098  https://www.samedfoundation.org/about/progress...
1099             https://www.missoulapubliclibrary.org/

[1100 rows x 1 columns]

In [50]:
legitimateURL.shape

(1100, 1)

Now we have our 5000 Legitimate and 5000 Phishing URLs ready.

# Feature Extraction

### 1 URL Based Features

In [11]:
# importing the required libraries

# !pip install tldextract

import re
import tldextract
from urllib.parse import urlparse
import requests
from bs4 import BeautifulSoup
import ssl
import socket

#### 1.1 Number of dots in URL

In [12]:
def no_of_dots(url):
    # Check for the number of dots in URL
    if(url.count(".") >= 4):
        return 1
    else:
        return 0

#### 1.2 Presence of @ symbol in URL

In [13]:
def specialSymbol(url):
    # Check for '@' symbol in URL
    if "@" in url:
      return 1
    else:
      return 0

#### 1.3 Length of the URL

In [14]:
def lengthOfURL(url):
    # Check for length of URL
    if len(url) >= 74:
        return 1
    else:
        return 0

#### 1.4 Suspicious words in URL

In [15]:
def suspiciousWords(url):
    # List of suspicious keywords
    keywords = [
        "security",
        "login",
        "signin",
        "bank",
        "account",
        "update",
        "include",
        "webs",
        "online",
    ]

    # Check for presence of the above keywords in URL
    for keyword in keywords:
        if keyword in url.lower():
            return 1
        else:
            return 0

#### 1.5 Prefix or Suffix "-" in Domain

In [16]:
# Check for some prefix or suffix keywords in brand name with dash symbol
def prefixSuffix(url):
  if '-' in urlparse(url).netloc:
    return 1
  else:
    return 0

#### 1.6 Count of HTTP in URL

In [17]:
def countOfHttp(url):
    if(url.count("http") > 1):
        return 1
    else:
        return 0

#### 1.7 Shortening Services

In [18]:
#listing shortening services
shortening_services = r"bit\.ly|goo\.gl|shorte\.st|go2l\.ink|x\.co|ow\.ly|t\.co|tinyurl|tr\.im|is\.gd|cli\.gs|" \
                      r"yfrog\.com|migre\.me|ff\.im|tiny\.cc|url4\.eu|twit\.ac|su\.pr|twurl\.nl|snipurl\.com|" \
                      r"short\.to|BudURL\.com|ping\.fm|post\.ly|Just\.as|bkite\.com|snipr\.com|fic\.kr|loopt\.us|" \
                      r"doiop\.com|short\.ie|kl\.am|wp\.me|rubyurl\.com|om\.ly|to\.ly|bit\.do|t\.co|lnkd\.in|db\.tt|" \
                      r"qr\.ae|adf\.ly|goo\.gl|bitly\.com|cur\.lv|tinyurl\.com|ow\.ly|bit\.ly|ity\.im|q\.gs|is\.gd|" \
                      r"po\.st|bc\.vc|twitthis\.com|u\.to|j\.mp|buzurl\.com|cutt\.us|u\.bb|yourls\.org|x\.co|" \
                      r"prettylinkpro\.com|scrnch\.me|filoops\.info|vzturl\.com|qr\.net|1url\.com|tweez\.me|v\.gd|" \
                      r"tr\.im|link\.zip\.net"
     

# Checking for Shortening Services in URL (Tiny_URL)
def shorteningService(url):
    match=re.search(shortening_services,url)
    if match:
        return 1
    else:
        return 0

#### 1.8 Data URI

In [19]:
def has_data_uri(url):
  try:
    # Retrieve the website's HTML content
    response = requests.get(url)
    html_content = response.text

    # Use a regular expression to search for data URIs
    pattern = re.compile(r'data:[^;]+;base64,')
    data_uris = pattern.findall(html_content)

    # Print out the data URIs found
    if data_uris:
        return 1
    else:
        return 0
  except:
    return 1


# has_data_uri("https://www.base64-image.de/")

#### 1.9 SSL certificate

In [20]:
def check_ssl_certificate(url):
    try:
        hostname = url.split('//')[1].split('/')[0]
        context = ssl.create_default_context()
        with socket.create_connection((hostname, 443)) as sock:
            with context.wrap_socket(sock, server_hostname=hostname) as ssock:
                cert = ssock.getpeercert()
                if cert:
                    return 0
                else:
                    return 1
    except Exception:
        return 1

# Test case 1: Valid SSL certificate
# print(check_ssl_certificate("https://www.google.com"))

# Test case 2: Invalid SSL certificate (expired)
# print(check_ssl_certificate("https://expired.badssl.com"))

# Test case 3: Invalid SSL certificate (self-signed)
# check_ssl_certificate("https://self-signed.badssl.com")

# Test case 4: Invalid SSL certificate (untrusted root)
# check_ssl_certificate("https://untrusted-root.badssl.com")

# Test case 5: Invalid URL (non-https)
# check_ssl_certificate("http://example.com")

# Test case 6: Invalid URL (invalid format)
# check_ssl_certificate("invalid-url")

### 2 Login Form Based Features

#### 2.1 Fake Login Form

In [21]:
def get_dom_tree(url):
    # Send a GET request to the URL
    response = requests.get(url)

    # Parse the HTML content of the response using BeautifulSoup
    dom_tree = BeautifulSoup(response.content, "html.parser")

    # Return the DOM tree as a BeautifulSoup object
    return dom_tree


def fake_login_form(url):

  try:
    dom_tree = get_dom_tree(url)

    # Initialize the output variable F9 to 0
    res = 0

    # Get all the form tags in the DOM tree
    forms = dom_tree.find_all("form")

    # Loop through each form tag
    for form in forms:
        # Get the value of the action attribute
        action = form.get("action")

        # If the action attribute is blank, # or javascript:void(0)), set F9 to 1
        if not action or action == "#" or action == "javascript:void(0)":
            res = 1
        # If the action attribute is in the form of "filename.php", set F9 to 1
        elif ".php" in action:
            res = 1
        # If the action attribute contains a foreign base domain, set F9 to 1 and break the loop
        elif "://" in action and not action.startswith(url):
            res = 1
            break

    # Return the output variable F9
    return res
  except:
    return 1

### 3 Hyperlink based Features

#### 3.1 Number of webpages

In [22]:
def countPages(url):
  try:
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    links = soup.find_all(['img', 'script', 'frame', 'input', 'link'])
    links += soup.find_all('a', href=True)
    page_count = len(links)
    return page_count
  except:
    return 100

#### 3.2 Missing Hyperlink

In [23]:
def missingHyperlink(url):
  try:
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    links = soup.find_all('a', href=True)
    
    if(len(links) > 0):
      return 0
    else:
      return 1
  except:
    return 1

#### 3.3 Foreign Hyperlinks

In [24]:
def count_foreign_links(url):
  try:
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    links = soup.find_all('a', href=True)
    foreign_links = 0
    for link in links:
        parsed_link = urlparse(link['href'])
        if parsed_link.netloc != '' and parsed_link.netloc != urlparse(url).netloc:
            foreign_links += 1
    total_links = len(links)
    ratio = foreign_links / total_links
    return int(ratio > 0.5)
  except:
    return 1

#### 3.4 Empty Hyperlinks

In [25]:
def count_empty_links(url):
  try:
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    links = soup.find_all('a')
    empty_links = []
    for link in links:
        href = link.get('href')
        if href in ['#', '#content', 'JavaScript::void(0)']:
            empty_links.append(link)

    ratio = len(empty_links) / len(links)
    if ratio > 0.34:
      return 1
    else:
      return 0
  except:
    return 1

#### 3.5 Hyperlink errors

In [26]:
def count_error_links(url):
  try:
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    links = soup.find_all('a')
    error_links = []
    for link in links:
      try:
        response = requests.get(link.get('href'))
        if response.status_code in [403, 404]:
          error_links.append(link)

      except requests.exceptions.RequestException as e:
        error_links.append(link)

    ratio = len(error_links) / len(links)
    
    if ratio > 0.3 and len(links) > 0:
      return 1
    else:
      return 0

  except:
    return 1

#### 3.6 Hyperlink redirection

In [27]:
def count_redirection_links(url):
  try:
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    links = soup.find_all('a')
    redirection_links = []
    for link in links:
        response = requests.get(link.get('href'))
        if response.status_code in [301, 302]:
            redirection_links.append(link)

    ratio = len(redirection_links) / len(links)

    if ratio > 0.3:
      return 1
    else:
      return 0
  except:
    return 1

### 4 Copied CSS features

#### 4.1 Copied CSS

In [28]:
def has_foreign_css(url):
  try:
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    css_links = soup.find_all('link', rel='stylesheet')
    for css_link in css_links:
        href = css_link.get('href')
        if href.startswith('http') and not href.startswith(url):
            return 1
    return 0

  except:
    return 1

### 5 Web Identity Based Features

#### 5.1 External Favicon

In [29]:
def check_favicon(url):
  try:
    # extract domain from url
    domain = urlparse(url).netloc
    
    # send GET request to fetch page content
    response = requests.get(url)
    
    # parse page content using BeautifulSoup
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # extract favicon link from page
    favicon_link = soup.find('link', rel='shortcut icon')
    
    # check if favicon link is present and get its href value
    if favicon_link is not None:
        favicon_href = favicon_link.get('href')
    else:
        # if favicon link is not present, assume it is the default favicon
        favicon_href = '/favicon.ico'
    
    # send GET request to fetch favicon
    favicon_url = domain + favicon_href
    response = requests.get("http://" + favicon_url)
    
    # check if favicon belongs to same domain as website
    if urlparse(response.url).netloc == domain:
        return 0
    else:
        return 1
    
  except:
    return 1

### 6 Domain Based Features

#### 6.1 Google Index

In [30]:
def google_index(url):
  try:
    google = "https://www.google.com/search?q=site:" + url + "&hl=en"
    response = requests.get(google, cookies={"CONSENT": "YES+1"})
    soup = BeautifulSoup(response.content, "html.parser")
    not_indexed = re.compile("did not match any documents")

    if soup(text=not_indexed):
      return 1
    else:
      return 0
  except:
    return 1

#### 6.2 http/https in domain

In [31]:
def httpDomain(url):
  domain = urlparse(url).netloc
  if 'https' in domain:
    return 1
  else:
    return 0

# Computing Features

In [32]:
def extractFeatures(url, label):
  features = []

  features.append(no_of_dots(url))
  features.append(specialSymbol(url))
  features.append(lengthOfURL(url))
  features.append(suspiciousWords(url))
  features.append(prefixSuffix(url))
  features.append(countOfHttp(url))
  features.append(shorteningService(url))
  features.append(has_data_uri(url))
  features.append(check_ssl_certificate(url))
  features.append(fake_login_form(url))
  features.append(countPages(url))
  features.append(missingHyperlink(url))
  features.append(count_foreign_links(url))
  features.append(count_empty_links(url))
  features.append(count_error_links(url))
  features.append(count_redirection_links(url))
  features.append(has_foreign_css(url))
  features.append(check_favicon(url))
  features.append(google_index(url))
  features.append(httpDomain(url))
  features.append(label)

  return features

### 1 Legitimate URLs

In [39]:
legitimateFeatures = []
label = 0

for i in range(0, 1100):
  url = legitimateURL['URLs'][i]
  legitimateFeatures.append(extractFeatures(url, label))


feature_names = ['no_of_dots', 
                 'specialSymbol', 
                 'lengthOfURL', 
                 'suspiciousWords', 
                 'prefixSuffix', 
                 'countOfHttp', 
                 'shorteningService', 
                 'has_data_uri', 
                 'check_ssl_certificate', 
                 'fake_login_form', 
                 'countPages', 
                 'missingHyperlink', 
                 'count_foreign_links', 
                 'count_empty_links', 
                 'count_error_links',
                 'count_redirection_links', 
                 'has_foreign_css', 
                 'check_favicon', 
                 'google_index', 
                 'httpDomain',
                 'label']

legitimate = pd.DataFrame(legitimateFeatures, columns = feature_names)
legitimate

c:\Users\harsh\AppData\Local\Programs\Python\Python310\lib\site-packages\bs4\builder\__init__.py:545: XMLParsedAsHTMLWarning: It looks like you're parsing an XML document using an HTML parser. If this really is an HTML document (maybe it's XHTML?), you can ignore or filter this warning. If it's XML, you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the lxml package installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.
  warnings.warn(
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.


no_of_dots  specialSymbol  lengthOfURL  suspiciousWords  prefixSuffix  \
0             0              0            0                0             0   
1             0              0            0                0             0   
2             1              0            1                0             0   
3             0              0            0                0             0   
4             0              0            0                0             0   
..          ...            ...          ...              ...           ...   
295           0              0            0                0             0   
296           0              0            0                0             0   
297           0              0            0                0             0   
298           0              0            0                0             0   
299           0              0            0                0             0   

     countOfHttp  shorteningService  has_data_uri  check_ssl_certificate  \
0              0                  0             0                      0   
1              0                  0             0                      0   
2              0                  0             1                      0   
3              0                  0             0                      0   
4              0                  0             0                      0   
..           ...                ...           ...                    ...   
295            0                  0             1                      0   
296            0                  0             0                      0   
297            0                  0             1                      1   
298            0                  0             1                      1   
299            0                  0             0                      0   

     fake_login_form  ...  missingHyperlink  count_foreign_links  \
0                  1  ...                 0                    0   
1                  0  ...                 0                    0   
2                  0  ...                 0                    0   
3                  1  ...                 0                    0   
4                  0  ...                 0                    0   
..               ...  ...               ...                  ...   
295                0  ...                 0                    0   
296                1  ...                 0                    0   
297                1  ...                 1                    1   
298                1  ...                 1                    1   
299                0  ...                 1                    1   

     count_empty_links  count_error_links  count_redirection_links  \
0                    0                  0                        1   
1                    0                  1                        1   
2                    0                  1                        1   
3                    0                  0                        1   
4                    0                  1                        1   
..                 ...                ...                      ...   
295                  0                  0                        1   
296                  0                  0                        0   
297                  1                  1                        1   
298                  1                  1                        1   
299                  1                  1                        1   

     has_foreign_css  check_favicon  google_index  httpDomain  label  
0                  1              0             0           0      0  
1                  1              1             0           0      0  
2                  0              0             0           0      0  
3                  1              0             0           0      0  
4                  0              0             0           0      0  
..               ...            ...           ...         ...    ...  


### 2 Phishing URLs

In [34]:
phishingFeatures = []
label = 1

for i in range(0, 1400):
  url = phishingURL['url'][i]
  phishingFeatures.append(extractFeatures(url, label))


feature_names = ['no_of_dots', 
                 'specialSymbol', 
                 'lengthOfURL', 
                 'suspiciousWords', 
                 'prefixSuffix', 
                 'countOfHttp', 
                 'shorteningService', 
                 'has_data_uri', 
                 'check_ssl_certificate', 
                 'fake_login_form', 
                 'countPages', 
                 'missingHyperlink', 
                 'count_foreign_links', 
                 'count_empty_links', 
                 'count_error_links',
                 'count_redirection_links', 
                 'has_foreign_css', 
                 'check_favicon', 
                 'google_index', 
                 'httpDomain',
                 'label']

phishing = pd.DataFrame(phishingFeatures, columns = feature_names)
phishing

In [58]:
phishing = pd.read_csv('phishing_final.csv')
phishing

no_of_dots  specialSymbol  lengthOfURL  suspiciousWords  prefixSuffix  \
0              0              0            0                0             0   
1              0              0            0                0             0   
2              0              0            0                0             0   
3              0              0            0                0             0   
4              0              0            0                0             0   
...          ...            ...          ...              ...           ...   
1395           0              0            0                0             1   
1396           0              0            0                0             1   
1397           0              0            0                0             0   
1398           0              0            0                0             1   
1399           0              0            0                0             1   

      countOfHttp  shorteningService  has_data_uri  check_ssl_certificate  \
0               0                  0             1                      1   
1               0                  0             1                      1   
2               0                  0             1                      1   
3               0                  0             1                      1   
4               0                  0             1                      1   
...           ...                ...           ...                    ...   
1395            0                  0             1                      1   
1396            0                  0             1                      1   
1397            0                  0             1                      1   
1398            0                  0             1                      1   
1399            0                  0             1                      1   

      fake_login_form  ...  missingHyperlink  count_foreign_links  \
0                   1  ...                 1                    1   
1                   1  ...                 1                    1   
2                   1  ...                 1                    1   
3                   1  ...                 1                    1   
4                   1  ...                 1                    1   
...               ...  ...               ...                  ...   
1395                1  ...                 1                    1   
1396                1  ...                 1                    1   
1397                1  ...                 1                    1   
1398                1  ...                 1                    1   
1399                1  ...                 1                    1   

      count_empty_links  count_error_links  count_redirection_links  \
0                     1                  1                        1   
1                     1                  1                        1   
2                     1                  1                        1   
3                     1                  1                        1   
4                     1                  1                        1   
...                 ...                ...                      ...   
1395                  1                  1                        1   
1396                  1                  1                        1   
1397                  1                  1                        1   
1398                  1                  1                        1   
1399                  1                  1                        1   

      has_foreign_css  check_favicon  google_index  httpDomain  label  
0                   1              1             1           0      1  
1                   1              1             1           0      1  
2                   1              1             1           0      1  
3                   1              1             1           0      1  
4                   1              1             1           0      1  
...               

In [53]:
legitimate = pd.read_csv("legitimate_final.csv")
legitimate

no_of_dots  specialSymbol  lengthOfURL  suspiciousWords  prefixSuffix  \
0              0              0            0                0             0   
1              0              0            0                0             0   
2              0              0            0                0             0   
3              0              0            0                0             0   
4              0              0            0                0             0   
...          ...            ...          ...              ...           ...   
1095           0              0            0                0             0   
1096           0              0            0                0             0   
1097           0              0            0                0             0   
1098           0              0            0                0             0   
1099           0              0            0                0             0   

      countOfHttp  shorteningService  has_data_uri  check_ssl_certificate  \
0               0                  0             0                      0   
1               0                  1             0                      0   
2               0                  0             0                      0   
3               0                  0             0                      0   
4               0                  0             0                      0   
...           ...                ...           ...                    ...   
1095            0                  0             1                      1   
1096            0                  1             1                      1   
1097            0                  0             1                      1   
1098            0                  0             1                      1   
1099            0                  0             1                      1   

      fake_login_form  ...  missingHyperlink  count_foreign_links  \
0                   1  ...                 0                    1   
1                   0  ...                 0                    0   
2                   0  ...                 0                    1   
3                   1  ...                 0                    0   
4                   0  ...                 0                    0   
...               ...  ...               ...                  ...   
1095                1  ...                 1                    1   
1096                1  ...                 1                    1   
1097                1  ...                 1                    1   
1098                1  ...                 1                    1   
1099                1  ...                 1                    1   

      count_empty_links  count_error_links  count_redirection_links  \
0                     0                  0                        0   
1                     0                  1                        1   
2                     0                  1                        1   
3                     0                  0                        1   
4                     0                  0                        1   
...                 ...                ...                      ...   
1095                  1                  1                        1   
1096                  1                  1                        1   
1097                  1                  1                        1   
1098                  1                  1                        1   
1099                  1                  1                        1   

      has_foreign_css  check_favicon  google_index  httpDomain  label  
0                   0              0             1           0      0  
1                   0              0             0           0      0  
2                   1              1             0           0      0  
3                   1              0             1           0      0  
4                   1              0             0           0      0  
...               

# Creating the final dataset

Merging the phishing data and legitimate data into one data frame

In [54]:
dataset = pd.concat([legitimate, phishing]).reset_index(drop = True)
dataset = dataset.sample(frac=1).reset_index(drop = True)
dataset

no_of_dots  specialSymbol  lengthOfURL  suspiciousWords  prefixSuffix  \
0              0              0            0                0             0   
1              0              0            0                0             0   
2              0              0            0                0             0   
3              0              0            0                0             0   
4              0              0            0                0             0   
...          ...            ...          ...              ...           ...   
2495           0              0            0                0             1   
2496           0              0            0                0             1   
2497           0              0            0                0             0   
2498           0              0            0                0             0   
2499           0              0            0                0             0   

      countOfHttp  shorteningService  has_data_uri  check_ssl_certificate  \
0               0                  0             0                      0   
1               0                  0             0                      0   
2               0                  0             1                      1   
3               0                  0             1                      1   
4               0                  0             0                      0   
...           ...                ...           ...                    ...   
2495            0                  0             0                      0   
2496            0                  0             1                      1   
2497            0                  0             1                      1   
2498            0                  0             0                      0   
2499            0                  0             0                      0   

      fake_login_form  ...  missingHyperlink  count_foreign_links  \
0                   0  ...                 1                    1   
1                   0  ...                 0                    0   
2                   1  ...                 1                    1   
3                   1  ...                 1                    1   
4                   1  ...                 0                    0   
...               ...  ...               ...                  ...   
2495                0  ...                 0                    1   
2496                1  ...                 1                    1   
2497                1  ...                 1                    1   
2498                0  ...                 0                    0   
2499                0  ...                 0                    0   

      count_empty_links  count_error_links  count_redirection_links  \
0                     1                  1                        1   
1                     0                  1                        1   
2                     1                  1                        1   
3                     1                  1                        1   
4                     0                  0                        1   
...                 ...                ...                      ...   
2495                  0                  0                        0   
2496                  1                  1                        1   
2497                  1                  1                        1   
2498                  0                  1                        1   
2499                  0                  0                        1   

      has_foreign_css  check_favicon  google_index  httpDomain  label  
0                   0              0             0           0      0  
1                   0              0             0           0      0  
2                   1              1             0           0      1  
3                   1              1             0           0      1  
4                   1              1             0           0      0  
...               

In [55]:
dataset.shape

(2500, 21)

Thus, we have a data set of 2500 URLs

In [57]:
dataset.to_csv('url_dataset.csv', index = False)